chore(tools): script to bump openapi spec dependency versions

aldousalvarez opened the pull request:
1. Adds a new script that can be executed in order to bump all release
versions of the open API specs. These are one of two main categories:
1.1. The top level .info.version property of the OpenAPI spec JSON object
1.2. Any schema component references embedded anywhere deep inside the
JSON object that we define our specs as, for example when a connector
plugin's endpoint returns a response object whose schema is defined in
another package's OpenAPI spec file such as the core-api package's.
2. Fixed a bug in the "prettier" npm script which was pointing to a non-
existent configuration file. This was necessary to be fixed in this commit
because tests had to be carried out whether the auto-formatting applied
by the version bumping script is consistent with the auto-formatting that
is applied when we run prettier via the CLI (there was some struggle here
to figure out that we have to pre-format the JSON string held in-memory
via JSON.stringify() first otherwise the formatted version would diverge
from what the CLI does to the .json files on-disk)
3. We had to exclude the tag called v2.0.0-alpha-prerelease because it
was incorrectly named unfortunately: it is considered newer than
v2.0.0-alpha.1 by the semver-parser package but it is actually older,
e.g. we've issued v2.0.0-alpha-prerelease before we issued v2.0.0-alpha.1
(which is our mistake and this is the workaround to rectify that after
the fact without having to force delete the tag that we've already
pushed to the upstream main branch).
This exclusion is applied by default which might turn out to be a mistake
later but right now it does not look like there would be an edge case
where we need to handle this in any different way.
4. The script is re-entrant and will save a thorough report of what it did
to temporary .json files whose names are date and timestamped within the
project root's ./build/ subfolder.
5. The target version to bump to can be specified explicitly via the
--target-version CLI parameter but if it not provided then will default
to whatever is the latest sem-ver compatibly named tag in the git log.

Example to specify target version explicitly:
`yarn tools:bump-openapi-spec-dep-versions --target-version=v3.2.1`

Example to let the script figure out what the target version should be:
`yarn tools:bump-openapi-spec-dep-versions`

Fixes https://github.com/hyperledger/cacti/issues/2206

Co-authored-by: Peter Somogyvari <peter.somogyvari@accenture.com>

Signed-off-by: aldousalvarez <aldousss.alvarez@gmail.com>
Signed-off-by: Peter Somogyvari <peter.somogyvari@accenture.com>

petermetz commented on tools/custom-checks/bump-openapi-spec-dep-version.ts:
```diff
@@ -0,0 +1,161 @@
+import { exec } from "child_process";
+import util from "util";
+
+interface paths {
```
@aldousalvarez Please follow naming conventions and have the interface start with `I` and also be CapitalCased in general.

petermetz commented on tools/custom-checks/bump-openapi-spec-dep-version.ts:
```diff
@@ -0,0 +1,161 @@
+    requestBody: unknown;
+  };
+}
+interface packageJson {
```
@aldousalvarez Please follow naming conventions and have the interface start with I and also be CapitalCased in general.

petermetz commented on tools/custom-checks/bump-openapi-spec-dep-version.ts:
```diff
@@ -0,0 +1,161 @@
+  const branchName = "main";
+  const latestVersion = execPromise(
+    `git describe --tags --abbrev=0 ${branchName}`,
+  ).then(({ stdout }) => {
```
@aldousalvarez Please use `await` unless there's a specific reason not to. It makes the code slighly more readable and avoids the first steps down onto callback hell.

petermetz commented on tools/custom-checks/bump-openapi-spec-dep-version.ts:
```diff
@@ -0,0 +1,161 @@
+    return latestTag;
+  });
+
+  const currentLatestVersion = await latestVersion;
```
@aldousalvarez It's probably the best to extract the entire logic of obtaining the latest git release tag into a function of it's own so that it could potentially be re-used elsewhere in the tooling code as well.

petermetz commented on tools/custom-checks/bump-openapi-spec-dep-version.ts:
```diff
@@ -0,0 +1,161 @@
+    console.log(
+      `${TAG} = = = = CHECKING TOTAL NUMBER OF REFERENCES = = = = \n`,
+    );
+    function countReferences(obj: any): number {
```
@aldousalvarez Pleae don't define nested functions like this and instead aim to have a flat structure for them. This make make them have to have more parameters but this is a good thing in the sense that it will make the functions be closer to pure functions.

petermetz commented on tools/custom-checks/bump-openapi-spec-dep-version.ts:
```diff
@@ -0,0 +1,161 @@
+      );
+    } else if (packageVersion != currentLatestVersion) {
+      pkgJson.info.version = currentLatestVersion;
+      await fs.writeFile(filePathAbs, JSON.stringify(pkgJson, null, 2));
```
@aldousalvarez When I ran the script, it produced a diff that changed the entire file due to whitespace changes. Let's figure out a way to have all the openapi.json's auto-formatted consistently with the same tool (prettier is my recommendation) and then you can apply that formatting after you've saved the JSON files and so the diff will only contain the lines that you actually changed, e.g. the vX.Y.Z declarations.

petermetz commented on tools/custom-checks/bump-openapi-spec-dep-version.ts:
```diff
@@ -0,0 +1,161 @@
+    const spec = await fs.readJSON(filePathAbs);
+    console.log("");
+    console.log(
+      `${TAG} = = = = CHECKING TOTAL NUMBER OF REFERENCES = = = = \n`,
```
@aldousalvarez Please clean up the log statements to just be regular sentences announcing either a) what's about to happen or b) what just happened. I get that making the log messages appear as headings of a document is helpful when you are debugging the code yourself during development, but then they lose their meaning when blended with all the other log messages from all the other components.

petermetz reviewed (changes requested):
@aldousalvarez Mostly LGTM but I have a few nits as usual. Please see the comments above for the details!

petermetz reviewed (changes requested):
@aldousalvarez FYI: I've pushed some changes that were easier to make then explain. Please make the following changes on top of what I had just pushed:
1. There are still nested functions, that point was never addressed.
2. Don't use regular expression search & replace to perform the version bumps. The files need to  be parsed and then the versions need to be updated based on the parsed data not via regex so that the replacements are deterministic and immune to bugs that will later stem from the content of the files changing in a way that will break the regex.

aldousalvarez commented:
@petermetz Thank you for the update, I have now addressed the changes that you have requested and I already fixed the nested functions and also I have implemented prettier as the one that will format the JSONs.

petermetz reviewed (approved):
LGTM
